### PR TITLE
*Fix: GLThread and other threads access sDownloaderMap in a thread-un…

### PR DESCRIPTION
**Pull request by pwang08**
GLThread and other threads(`js_load_remote_image` creates CCDownloader in another thread) access `sDownloaderMap` in a thread-unsafe way, which will result in crash in some occasion on Android.

### CODE REVIEW

**Comment by dumganhar** [comment]
Hi, pwang08,
I notice that `Downloader` instance is usually constructed in GL thread.
And the callback functions are also invoked from GL thread. Could this cause thread problems?

``` java
Cocos2dxHelper.runOnGLThread(new Runnable() {
            @Override
            public void run() {
                nativeOnProgress(_id, id, downloadBytes, downloadNow, downloadTotal);
            }
        });
```

**Comment by pwang08** [comment]
`void __JSDownloaderDelegator::downloadAsync()
{
    retain();
    auto t = std::thread(&__JSDownloaderDelegator::startDownload, this);
    t.detach();
}
`

As the code above shows: `__JSDownloaderDelegator::startDownload` is called in another thread, in which a `Downloader` instance is created. As the code below shows:

`
void __JSDownloaderDelegator::startDownload()
{
    if (auto texture = Director::getInstance()->getTextureCache()->getTextureForKey(_url))
    {
        onSuccess(texture);
    }
    else
    {
        _downloader = std::make_shared<cocos2d::network::Downloader>();
    }
}
`

**Comment by dumganhar** [comment]
@pwang08 , it makes sense.

@pandamicro , Could you please check it since if I remember correctly, __JSDownloaderDelegator stuff was written by u.

**Comment by pandamicro** [comment]
Yes, it was originally created by me, at the time, there was a `performFunctionInCocosThread` to ensure everything is thread safe. But @yangws  have refactored the network module and removed these code because network it self will make sure callbacks are invoked in cocos thread.

**Comment by minggo** [comment]
just close/reopen it to trigger ci. This PR seems good.
